### PR TITLE
Document coverage export script for assembling pretraining dataset

### PR DESCRIPTION
**Pull request by nadahlberg**
@mlissner here’s the script for pulling a case index with some useful document coverage stats for assembling a training set. Is there a better place to put this? I added it to the search app’s management commands but that doesn’t seem right.

Just to be safe I added an elastic and a psql version (see flag —psql), but the psql one is preferable since it will also pull in data about the rss coverage. 

The general idea is that by merging this data with the bulk docket header data we can find a high quality sample of cases where: (i) most or all documents are available for a given case, bolstering the representation of less popular document types, (ii) cases are nicely distributed across various case types, courts, nature suit, etc. (iii) a balance of long / short cases, and (iv) mostly non-rss dockets, while making sure that we have a nice distribution for the rss entries we do include. I’d also like to add a generous portion of opinions and maybe even raw html dockets (billing table redacted).

If y’all can run this and share the export I can run some tests and get some initial token count / training cost estimates so we can decide how big we want to go. Thanks!

### CODE REVIEW

File: cl/search/management/commands/export_doc_coverage.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            if not options["overwrite"]:
+                raise FileExistsError(
+                    f"File already exists at {options['out_path']}. Use --overwrite to overwrite."
+                )
+            out_path.unlink()
+
+        handler = self.handle_psql if options["psql"] else self.handle_es
+
+        min_year = Docket.objects.earliest("date_filed").date_filed.year
+        max_year = Docket.objects.latest("date_filed").date_filed.year
```
Comment (mlissner) [inline cl/search/management/commands/export_doc_coverage.py:54]: This can just be now, right?

File: cl/search/management/commands/export_doc_coverage.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        logger.info(f"Processing dockets from {min_year} to {max_year}...")
+        for year in tqdm(
```
Comment (mlissner) [inline cl/search/management/commands/export_doc_coverage.py:57]: Can we add a configurable sleep in our for loop, perhaps? I could see this being pretty nasty on some of our biggest tables, so at least putting a sleep in it will spread the load a bit. (Should probably do it for ES too.)

Comment (mlissner) [reply cl/search/management/commands/export_doc_coverage.py:57]: We should add a log-line too, so we know what kind of progress we're making each time we go through the loop, otherwise, we won't know if one query is taking five minutes or if the whole thing is.

File: cl/search/management/commands/export_doc_coverage.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        if out_path.exists():
+            if not options["overwrite"]:
+                raise FileExistsError(
+                    f"File already exists at {options['out_path']}. Use --overwrite to overwrite."
+                )
+            out_path.unlink()
+
+        handler = self.handle_psql if options["psql"] else self.handle_es
+
+        min_year = Docket.objects.earliest("date_filed").date_filed.year
```
Comment (mlissner) [inline cl/search/management/commands/export_doc_coverage.py:53]: We use dockets both in case law and in RECAP, so you'll want to make sure that we only include RECAP data here and below. That can be done via the `source` field, I believe (there are examples around), or, probably, by checking for dockets that have entries.

We have millions of docket objects with no entries, if that might throw you off.

**Review by mlissner — COMMENTED** [review]
Looks pretty OK to me, and I think the file location is fine. A couple little notes. My biggest concern is that the aggregation queries are going to be *rough* on the DB.

I'm inclined to do the postgres approach and haven't reviewed the ES one, but if we go the ES route, we can probably use estimated queries with an accuracy threshold, like we do on the front end, and that might help with performance.

File: cl/search/management/commands/export_doc_coverage.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            "--overwrite",
+            action="store_true",
+            help="Overwrite the output file if it exists",
+        )
+
+    def handle(self, *args, **options):
+        out_path = Path(options["out_path"])
+        if out_path.exists():
+            if not options["overwrite"]:
+                raise FileExistsError(
```
[the remark this answers is not in the document]
Comment (nadahlberg) [reply cl/search/management/commands/export_doc_coverage.py:60]: It's a built in. Make it explicit?

File: cl/search/management/commands/export_doc_coverage.py (hunk 3 above)
Comment (nadahlberg) [reply cl/search/management/commands/export_doc_coverage.py:67]: Is it this? `source__in=Docket.RECAP_SOURCES()`

**Comment by nadahlberg** [comment]
Added track_total_hits=False for the ES query (maybe redundant with size=0). And a max-es-buckets configuration and dropped the default to half a million, hopefully that will reduce the load there and still give us more than we need per year.

Would it be better to batch by month or something more granular?

(thank you formatting bot, forgot to rerun black at the last second)

File: cl/search/management/commands/export_doc_coverage.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/search/management/commands/export_doc_coverage.py:60]: Oh. My bad! I thought I knew all the Python exceptions. Guess not.

File: cl/search/management/commands/export_doc_coverage.py (hunk 3 above)
Comment (mlissner) [reply cl/search/management/commands/export_doc_coverage.py:67]: Yeah, that looks right!

**Comment by mlissner** [comment]
We'll find out about the per-year or per-month business eventually, I guess, but we can start per year and see what happens. 

> Added track_total_hits=False for the ES query (maybe redundant with size=0). And a max-es-buckets configuration and dropped the default to half a million, hopefully that will reduce the load there and still give us more than we need per year.

Regrettably, I don't know much about this, but @albertisfu will certainly have opinions. Alberto, can you take a quick look at the code for querying ES and see if you approve of it generally? This doesn't have to be perfect, it's just for generating stats, but we at least don't want it to crash things!

**Comment by nadahlberg** [comment]
Yikes, I ran a similar query on a larger dataset and realized terms aggs cap out at 10,000 buckets. I also tried a scripted metric, but the response sizes were too large on any meaningful amount of data. 

This is the same thing, but wraps it in a composite agg so that it gets done in minibatches. Will probably be slow but hopefully gentle on memory. I think the composite agg is more efficient than just scrolling through all the data, but not totally sure. A little out of my depth here 😅. Thanks in advance for any feedback @albertisfu !

**Review by albertisfu — COMMENTED** [review]
This looks great @nadahlberg. I just left a few small comments.

Additionally I have a couple of questions:

You mentioned the `max-es-buckets` configuration in a comment, but I didn't see it in the code. Does that mean the command will retrieve data from all the dockets in Elastic and the database? If so, I think that could lead to a memory issue, considering the command is designed to request all data for a given year before storing it in the CSV file.

We have years with up to 4 million dockets, and since each aggregation request for 10,000 dockets is around 4MB (I tested it), that would result in approximately 400 requests, each with 4MB, totaling around 1,600MB of data in memory before writing to the CSV.

I wonder if this is ok for the instance where this command will run, @mlissner, or if it would be better to implement some memory optimizations, perhaps writing each response directly to the CSV?
 
 In a comment Mike mentioned:
>  We'll find out about the per-year or per-month business eventually, I guess, but we can start per year and see what happens.
 
Is this intended only for batch processing to happen on a yearly basis? I noticed that even though batches are processed per year, they end up in the same CSV file. Is that the expected behavior? I had imagined having a separate file for each year's data.

**Comment by albertisfu** [comment]
> I thought the script was just going to put stats into the CSV for now, not cases, so that shouldn't be a memory issue?

Yeah, that's right! However, my concern is related to fetching ES data for an entire year. Since all the data for a given year is stored in memory before being appended to the CSV, it might lead to high memory usage.

For instance, this ES response for the request is 3.9MB, and we could have up to 400 such responses in memory for a single year.

[response_agg.json](https://github.com/user-attachments/files/18917461/response_agg.json)

I see that the JSON is being simplified into a dictionary with only a few values, so perhaps the actual memory usage will be lower than the 1,600MB I previously mentioned.

**Comment by mlissner** [comment]
Alberto, back to you for another review. This is more or less a one-off script, so if it seems safe and effective, it doesn't have to be perfect. Thanks!

**Comment by mlissner** [comment]
Lovely. Merged!

**Comment by mlissner** [comment]
Currently running this with: 

```
./manage.py export_doc_coverage --sleep 1 --out-path /storage/doc-coverage/coverage.csv
```

That completed overnight and I wormhole'd the file to my laptop. Will give to @nadahlberg today.
